sdks/java/io/csv/src/main/java/org/apache/beam/sdk/io/csv/CsvIO.java

```diff
@@ -0,0 +1,384 @@
+    abstract @Nullable String getFilenamePrefix();
+
+    abstract CSVFormat getCSVFormat();
+
+    abstract @Nullable String getPreamble();
+
+    abstract @Nullable Compression getCompression();
+
+    abstract @Nullable Integer getNumShards();
+
+    abstract String getFilenameSuffix();
+
+    abstract @Nullable ResourceId getTempDirectory();
+
+    abstract @Nullable List<String> getSchemaFields();
+
+    abstract Builder<T> toBuilder();
+
+    @AutoValue.Builder
+    abstract static class Builder<T> {
+
+      abstract Builder<T> setFilenamePrefix(String value);
+
+      abstract Builder<T> setCSVFormat(CSVFormat value);
+
+      abstract Optional<CSVFormat> getCSVFormat();
+
+      abstract Builder<T> setPreamble(String value);
+
+      abstract Builder<T> setCompression(Compression value);
+
+      abstract Builder<T> setNumShards(Integer value);
+
+      abstract Builder<T> setFilenameSuffix(String value);
+
+      abstract Optional<String> getFilenameSuffix();
+
+      abstract Builder<T> setTempDirectory(ResourceId value);
+
+      abstract Builder<T> setSchemaFields(List<String> value);
+
+      abstract Write<T> autoBuild();
+
+      final Write<T> build() {
+        if (!getCSVFormat().isPresent()) {
+          setCSVFormat(CSVFormat.DEFAULT);
+        }
+
+        if (!getFilenameSuffix().isPresent()) {
+          setFilenameSuffix(DEFAULT_FILENAME_SUFFIX);
+        }
+
+        return autoBuild();
+      }
+    }
+
+    @Override
+    public void populateDisplayData(DisplayData.Builder builder) {
+      if (getFilenamePrefix() != null) {
+        builder.add(DisplayData.item("filenamePrefix", getFilenamePrefix()));
+      }
+      if (getCSVFormat() != null) {
+        builder.add(DisplayData.item("csvFormat", getCSVFormat().toString()));
+      }
+      if (getPreamble() != null) {
+        builder.add(DisplayData.item("preamble", getPreamble()));
+      }
+      if (getCompression() != null) {
+        builder.add(DisplayData.item("compression", getCompression().name()));
+      }
+      if (getNumShards() != null) {
+        builder.add(DisplayData.item("numShards", getNumShards()));
+      }
+
+      builder.add(DisplayData.item("filenameSuffix", getFilenameSuffix()));
+
+      if (getTempDirectory() != null) {
+        builder.add(DisplayData.item("tempDirectory", getTempDirectory().getFilename()));
+      }
+    }
+
+    @Override
+    public PDone expand(PCollection<T> input) {
+      if (!input.hasSchema()) {
+        throw new IllegalArgumentException(
+            String.format(
+                "%s requires an input Schema. Note that only Row or user classes are supported. Consider using TextIO or FileIO directly when writing primitive types",
+                Write.class.getName()));
+      }
+
+      Schema schema = input.getSchema();
+      SerializableFunction<T, Row> toRowFn = input.getToRowFunction();
+      PCollection<Row> rows =
+          input.apply(MapElements.into(rows()).via(toRowFn)).setRowSchema(schema);
```

Comment:
It is always good to give names for each PTransform that is applied so that it is easier to understand what is occurring.

```suggestion
          input.apply("To Rows", MapElements.into(rows()).via(toRowFn)).setRowSchema(schema);
```